### PR TITLE
Add Call Graph [WIP]

### PR DESCRIPTION
**Pull request by Amejonah1200**
recaf

TODO:
- [x] Call Graph Docking
  - [x] Move to own Window
  - [x] Bundle in one TabPane for Docking
  - [x] Refocus of method (navigation)
- [x] Method Handles (?)
- [x] Dynamic methods
- [x] Fix NPE
- [x] Support for workspace updates:
  - [x] Add Library
  - [x] Add new class
  - [x] Update Class
  - [x] Delete Class
  - [x] Remove Library
- [x] LinkerResolver caching
- [x] Show unresolved calls
- [x] Threading Issues?
- [x] Cache ClassInfo?
- [ ] Optimize renaming and other changes that affect this feature
- [x] ~~Make Toggle for Call Graph~~

### CODE REVIEW

File: recaf-ui/src/main/java/me/coley/recaf/ui/pane/MethodCallGraphPane.java [hunk 1]
```diff
@@ -172,7 +172,7 @@ protected void updateItem(MethodInfo item, boolean empty) {
 					removeEventFilter(MouseEvent.MOUSE_PRESSED, onClickFilter);
 			} else {
 				onClickFilter = null;
-				Text classText = new Text(TextDisplayUtil.escapeShortenPath(item.getOwner()));
+				Text classText = new Text(EscapeUtil.escape(item.getOwner()));
```
Comment (Col-E) [inline recaf-ui/src/main/java/me/coley/recaf/ui/pane/MethodCallGraphPane.java:175]: Was renamed to `shortenEscapeLimit`.

Should use `TextDisplayUtil.shortenEscapeLimit(...)` for consistency of using `TextDisplayUtil` when possible _(and limiting for stupid long names from obfuscators)_

**Comment by Amejonah1200** [comment]
This PR is stuck, as it needs an inheritance graph which can resolve methods correctly compliant with JVM behavior.
https://docs.oracle.com/javase/specs/jvms/se8/html/jvms-5.html#jvms-5.4.3.3

**Comment by Amejonah1200** [comment]
Thanks to @xxDark with https://jitpack.io/#xxDark/jlinker , I will try to implement further calls.

**Comment by Amejonah1200** [comment]
On another note, I will probably swutch to a standalone tab (dock) instead of it being at the right.

File: recaf-ui/src/main/java/me/coley/recaf/ui/context/MethodContextBuilder.java [hunk 2]
```diff
@@ -77,6 +81,7 @@ public ContextMenu build() {
 			menu.getItems().add(refactor);
 			Menu view = menu("menu.view", Icons.EYE);
 			view.getItems().add(action("menu.view.methodcfg", Icons.CHILDREN, this::graph));
```
Comment (Col-E) [inline recaf-ui/src/main/java/me/coley/recaf/ui/context/MethodContextBuilder.java:83]: Merge note to self: Can make this more specific/clear now that there are two _"graph"_ options

**Comment by Col-E** [comment]
Migrated to new branch under: https://github.com/Col-E/Recaf/tree/dev4/recaf-api/src/main/java/software/coley/recaf/services/callgraph

Anything seem to be missing?

**Comment by Col-E** [comment]
image

Feedback from xDark, can address on `dev4`

**Comment by Amejonah1200** [comment]
image
There are reports, these colors are not very suitable. If someone can give better ones, I would gladly accept and integrate.

Otherwise, I would put this as ready to be merged.

**Comment by Amejonah1200** [comment]
image

**Comment by Col-E** [comment]
Its finally been ported over to dev4 🥳 

06987b0996c417dd17e8b743c1423d1122e0644d

**Comment by Col-E** [comment]
Closing now that 4.X has the same functionality, we can continue from there rather than on this 3.X branch

**Pull request closed without merging**